### ISSUE TITLE
contentSecurityPolicy blocking cdn

### ISSUE THREAD

**Issue by krishnaTORQUE** [helmetjs/helmet#274]
when I am using `app.use(helmet())` its blocking all other cdn & static urls
but if I use the code below then cdn & static urls working fine.
```
app.use(
  helmet({
    contentSecurityPolicy: false,
  })
)
```

**Comment by EvanHahn** [comment]
This is probably because you need to configure the Content Security Policy to allow your CDNs. It's hard to give an example without knowing your specific use case, but the errors in your browser's console should be helpful.

You can also disable CSP, like you've done here, but then you lose that security benefit.

**Comment by krishnaTORQUE** [comment]
Thank you for your quick reply.
Yes you are absolutely correct that `disable CSP, you lose that security benefit`, thats why I open that issue.

`Refused to load the script 'https://static-url.com' because it violates the following Content Security Policy directive: "script-src 'self'". Note that 'script-src-elem' was not explicitly set, so 'script-src' is used as a fallback`

this is error i am getting in console. For Static Url & also google adsense and other all the external urls.

can you please tell me how can I `configure the Content Security Policy to allow your CDNs` ?

**Comment by krishnaTORQUE** [comment]
I have added this code below but still getting the error

```
app.use(
  helmet({
    contentSecurityPolicy: {
      directives: {
        defaultSrc: ['"self"'],
        scriptSrc: [
          '"self"',
          'localhost',
          'google-analytics.com',
          'pagead2.googlesyndication.com',
        ],
      },
    },
  })
)
```

**Comment by krishnaTORQUE** [comment]
Another try with this code below & still the same error
```
app.use(
  helmet({
    contentSecurityPolicy: {
      directives: {
        'default-src': ["'self'"],
        'base-uri': ["'self'"],
        'style-src': ["'self'", 'localhost'],
        'script-src': [
          "'self'",
          'localhost',
          'www.google-analytics.com',
          'pagead2.googlesyndication.com'
        ],
      },
    },
  })
)
```

**Comment by krishnaTORQUE** [comment]
it is also blocking api routes with 404 in production

**Comment by krishnaTORQUE** [comment]
after tons of search I found it working with little issue.

```
app.use(
  helmet({
    contentSecurityPolicy: {
      directives: {
        defaultSrc: [
          "'self'",
          "'unsafe-inline'",
          'localhost',
          '*.main-domain.com',
          '*.google.com',
          '*.google.co.in',
          '*.google-analytics.com',
          '*.googlesyndication.com',
          '*.googleapis.com',
          '*.doubleclick.net',
        ],
      },
    },
  })
)
```
1. still `svg` images not working
2. I have added `google.com` & `google.co.in` but is there any way use `*.google.*` I found it is not working

**Comment by krishnaTORQUE** [comment]
alright so my final code is below which solved the svg issue too.

```
app.use(
  helmet({
    contentSecurityPolicy: {
      directives: {
        defaultSrc: [
          "'self'",
          "'unsafe-inline'",
          'data:',
          'localhost',
          'main-domain.com',
          '*.main-domain.com',
          '*.google.com',
          '*.google.co.in',
          '*.google-analytics.com',
          '*.googlesyndication.com',
          '*.googleadservices.com',
          '*.googletagservices.com',
          '*.googleapis.com',
          '*.doubleclick.net',
          '*.gstatic.com',
          'youtu.be',
          '*.youtu.be',
          '*.youtube.com',
        ],
      },
    },
  })
)
```
the only issue i am facing is in production. api send all request 404
example api path `'/api/user`'

another issue 
I have added `google.com` & `google.co.in` but is there any way use *.google.* I found it is not working

**Comment by EvanHahn** [comment]
What you've done should mostly work, but keep in mind that `default-src` is the fallback for lots of different things, and may allow things you don't want.

For example, if you're using JavaScript from `example.com` but not using images, you might do something like this instead of using `default-src`:

```js
// ...
scriptSrc: [
  'example.com'
],
// ...
```

If you use `default-src` instead, you'll be able to load JavaScript, CSS, images, WebSockets, iframes, and more from these domains, which is probably not what you want. You may prefer to use a short `default-src` with longer values for `script-src`, `style-src`, and so on.

Using `*.google` as a directive value should work. What error do you see?

I'm going to close this issue because it doesn't indicate a problem with Helmet, but I'm happy to continue discussion here.

**Comment by krishnaTORQUE** [comment]
```
Refused to load the image 'https://www.google.com/ads/ga-audiences?t=sr&aip=1&_r=4&slf_rd=1&v=1&_v=j86…1-1&cid=212716265.1603472304&jid=2120763442&_u=AACAAEAAAAAAAC~&z=937627603' because it violates the following Content Security Policy directive: "img-src 'self' localhost data: *.google *.google-analytics.com".
```
this is the error I am getting

**Comment by krishnaTORQUE** [comment]
also all api requests return 404

**Comment by EvanHahn** [comment]
To break down the error:

    Refused to load the image 'https://www.google.com/ads/ga-audiences?t=sr&aip=1&_r=4&slf_rd=1&v=1&_v=j86…1-1&cid=212716265.1603472304&jid=2120763442&_u=AACAAEAAAAAAAC~&z=937627603' because it violates the following Content Security Policy directive: "img-src 'self' localhost data: *.google *.google-analytics.com".

It's refusing to load an image from `https://www.google.com/ads/...` because it violates the `img-src` directive. It looks like the `img-src` directive looks like this:

    img-src 'self' localhost data: *.google *.google-analytics.com

This should look similar to your JavaScript code. You'll want to add something like `google.com` to the array.

Does that help?

**Comment by krishnaTORQUE** [comment]
yes if I am adding `google.com` then its working fine
is there any way to add google domain like we are doing here `*.google.com` similar to `*.google.*` ?

**Comment by EvanHahn** [comment]
I'm not sure, but I wouldn't. If someone serves malicious code on `www.google.evil.com`, for example, you'd be vulnerable.

**Comment by EvanHahn** [comment]
API 404s shouldn't be related to CSP. Are you seeing CSP errors?

**Comment by krishnaTORQUE** [comment]
> API 404s shouldn't be related to CSP. Are you seeing CSP errors?

i am not sure its from CSP or not but when ever I enable `app.use(helmet())` I get 404 from api

**Comment by EvanHahn** [comment]
I don't know why that is happening. Some ideas:

- Make a new app to try to reproduce the problem
- Ask on Stack Overflow
- Try to make requests to similar routes to try to figure out what part is failing

Hope this helps.

**Comment by kerimkaan** [comment]
Hi guys,

I've an issue like this. Frontend (React, Hosted on Vercel) sended a request to my backend (REST API, Private server) and fetch failed. I got this console message:

> Access to fetch at 'https://BACKEND' from origin 'https://FRONTEND.vercel.app' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

Result:
`TypeError: Failed to fetch`

Backend and frontend support HTTPS. My CSP directives include defaultSrc: self, *.vercel.app ; scriptSrc, scriptElemSrc, styleSrc, fontSrc same as defaulSrc. I've tried "frameAncestors" directive for CSP but failed again. Error message:

> Access to fetch at 'https://BACKEND/' from origin 'https://FRONTEND.vercel.app' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

I've some ideas: My route (`POST` /auth) can take request, but Vercel is sending different method `OPTIONS`. I think, Helmet can support [Access-Control-Request-Method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS#Preflighted_requests_in_CORS) but if you refer to #79, i will disable cors. Turn the `cors` package. 

What can i do? Should I disable CORS?

**Comment by EvanHahn** [comment]
@kerimkaan It looks like you'll need to enable CORS on the backend. Helmet doesn't deal with CORS at all, so you may wish to [read about CORS on MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) or use the [cors package](https://www.npmjs.com/package/cors).

**Comment by chrisdel101** [comment]
I'm running into this too `Content Security Policy: The page’s settings blocked...` on a `https://cdn.buymeacoffee.com` call. 

I've tried tons of things but none work, including `contentSecurityPolicy: false`, and trying to put that URL into the `defaultSrc` or `scriptSrc` fields. (like in docs). Where is the correct place to add the URL it should allow? I'm not too familiar with these fields.

Disabling Helmet resolves the issue.

**Comment by EvanHahn** [comment]
@chrisdel101 Two questions:

1. What is the value of the Content-Security-Policy header? You can find this in the Network tab of your browser's developer tools.
1. Could you create a small sample app that reproduces your problem?

**Comment by chrisdel101** [comment]
It looks like `contentSecurityPolicy: false` _does_ work but was blocked by a browser cache issue.  If all else fails I'll use this.


More refined, my setup is like this:
```
contentSecurityPolicy: {
      directives: {
        defaultSrc: ["'self'", "https://cdn.buymeacoffee.com/*", "http://placekitten.com/*"]
      }
    }
```
The header looks like this:
```
default-src 'self' https://cdn.buymeacoffee.com/* http://placekitten.com/*;base-uri 'self';font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests
```
Other http calls are working okay, even though they are not included here. So not sure what the issue is.
